src/Template.php

```diff
@@ -0,0 +1,115 @@
+{
+    /**
+     * Templates directory contains all application templates.
+     * @var string
+     */
+    private $templatesDir;
+
+    /**
+     * Default context of template engine (usage of $this in template belongs to
+     * set context).
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * Pool of registered functions in the application.
+     * @var array
+     */
+    private $functions = [];
+
+    /**
+     * Pool of global variables.
+     * @var array
+     */
+    private $variables = [];
+
+    /**
+     * Class constructor.
+     */
+    public function __construct(string $templatesDir, Context $context)
+    {
+        $this->templatesDir = $templatesDir;
+        $this->context = $context;
+    }
+
+    /**
+     * This enables adding new variables in the template scope right after
+     * initializing a template engine. Some variables in templates are like
+     * parameters or globals and should be added only on one place instead of
+     * repeating them at each ...->render() call.
+     */
+    public function add(array $vars = []): void
+    {
+        $this->variables = array_merge($this->variables, $vars);
+    }
+
+    /**
+     * Add a template helper function as a callable defined in the (front)
+     * controller to the template scope. Useful when a custom function or class
+     * method need to be called in the template. A wrapper around the
+     * Context::addFunction().
+     */
+    public function addFunction(string $name, callable $callback): void
+    {
+        $this->context->addFunction($name, $callback);
+    }
+
+    /**
+     * Renders given template file and populates its scope with variables
+     * provided as array elements. Each array key is a variable name in template
+     * scope and array item value is set as a variable value. To not mess with
+     * the variable scopes too much closure arguments are retrieved dynamically
+     * via the func_get_arg(). Note that $this pseudo-variable in the closure
+     * refers to the Context::class scope.
+     */
+    public function render(string $template, array $vars = []): string
+    {
+        $closure = \Closure::bind(
+            function() {
+                ob_start();
+                extract(func_get_arg(2), EXTR_SKIP);
+
+                try {
+                    include func_get_arg(0).'/'.func_get_arg(1);
+                } catch (\Exception $e) {
+                    ob_end_clean();
+                    throw $e;
+                }
+
+                ob_end_clean();
+
+                ob_start();
+                extract($this->layoutVars);
+
+                include func_get_arg(0).'/'.$this->layout;
+
+                return ob_get_clean();
+            },
+            $this->context,
+            Context::class
+        );
+
+        // Merge variables set on the template render level and ones defined
+        // earlier in the engine creation level.
+        $vars = array_merge($this->variables, $vars);
```

Comment:
array_merge does behave different when an array with integer indexes are merged to a array that contains values for the same integer index. Use `array_replace` instead.
```suggestion
        $vars = array_replace($this->variables, $vars);
```

Reply:
@nhlm That won't be an issue as to the above code. $1 is _technically_ not a valid variable name  combined with the EXTR_SKIP flag (Okay you can cheese it with ${'1'}).

Reply:
It does not matter what the code afterwards does. Logic failures are logic failures. Better removing issues now, as seeking for them later. So, yes, it would be an issue, a maintaining issue.

Reply:
Requiring that the assignments should *always* be replace but **never** *change the type of a assigned value or behave different* is what you want. If you use strategies who behave different is a logic failure by definition. Logic failures are not influenced by the decisions made outside of the method who defines the logic. It is not about solving a problem, it is about preventing future problems. You do aim to fulfill **YAGNI**, but problem prevention when it has close to zero costs because it requires only to switch to a different function that **always fulfills what is expected** is not violating **YAGNI**, it is just implementing expected behaviors. If i would suggest to check the array for numeric indexes, it wouldn't be needed because the array items who rises conflicts then, wouldn't ever reach the scope, because no prefixing is configured, if it would be it would be a strategy change, not a logical failure.

Reply:
As a general rule of thumb, I reckon it's better to prevent a problem in the first place, rather than risk needing to solve it later, if it can be prevented in the first place. I mean.. In cases like these (`array_merge` versus `array_replace`), it might seem like a very small thing to bother about, but then, if the cost of implementation is equally small (e.g., doesn't cost much time, involves just changing one function call, one line of code, whatever), and a change to the codebase is in progress anyway (e.g., this PR), I would generally just go ahead and implement it (assuming no foreseeable negative consequences, of course).

Reply:
That's a fair argument and I can accept that, as perhaps I interpreted your initial comment differently.

Reply:
I think that's very nice catch and suggestion. `array_replace()` has been used instead of `array_merge()`. Yes, now also additionally when this happens:
```php
$template->render('pages/page.php', [
'Some value without key'
]);
```

The Exception is thrown instead. It happens too many times when developing the templates actually, yes.